### ISSUE TITLE
Azure Functions v3 - Client Access Tokens

### ISSUE THREAD

**Issue by mpaul31** [IdentityModel/IdentityModel.AspNetCore#110]
Hello,

I am having trouble following the v3 samples and configuring my function app to use the client access token handler.

The `FunctionsStartup.Configure` method only provides a reference to the `IServiceCollection` so I am not able to call `app.UseAuthorization`.

Is there something I am missing or another way around this?

Thanks!

**Comment by leastprivilege** [comment]
I have never used a function app. So I assume it is not made for that.

**Comment by Vidarls** [comment]
Ok, So I've spent this morning banging my head against this as well:

## Tl;Dr:

The code here gets you a working Client Credentials Http Client using `IdentityModel.AspNetCore` 

## Longer story

My Goal: Just get a client credentials http client with access token management up and running in an Azure Function triggered by a Queue messsage (no HTTP request).

1. @mpaul31 with a function app, it probably makes more sense to work from the Worker sample (https://github.com/IdentityModel/IdentityModel.AspNetCore/tree/master/samples/WorkerService) as that leaves out the requirement to set up authentication handlers. (Those are not really needed for a stripped down client credentials scenario anyway,
2. Due to Function apps not being quite and Asp.net core web app, but sort of anyway => The default config in `https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/master/src/AccessTokenManagement/TokenManagementServiceCollectionExtensions.cs` causes error messages: Complaining about authentication handlers not being registered.
3. Fix: Create a slightly fragile configuration where only the minimum of services are registered avoiding triggering of the Authentication strangeness:

## The config that works

```c#
using IdentityModel.AspNetCore.AccessTokenManagement;
using Microsoft.AspNetCore.Authentication;
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Options;

// Configure dependency injection for Azure functions
[assembly: FunctionsStartup(typeof(Functions.Startup))]

namespace Functions
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            // Configure HttpClient with automatic access token
            // management
            // Can not use the default one liners as they depend on specific 
            // asp.net stuff not available in function apps at the moment.

            // The following section is a re-implementation of 
            // builder.Services.AddAccessTokenManagement(options => (configure options))
            // src: https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/master/src/AccessTokenManagement/TokenManagementServiceCollectionExtensions.cs
            // Difference: Only options and services related to a client credentials flow is configured
            //             AND: it only supports 1! client configuration. (Possible workaround is to always used named clients)
            //             Leaving the default config caused error messages
            //             related to Authentication handlers and schemes (that we are not
            //             using in our function app.
            builder.Services.Configure<AccessTokenManagementOptions>(o => 
            {
                o.Client.Clients.Add("identityserver", new IdentityModel.Client.ClientCredentialsTokenRequest
                {
                    Address = "http://myIdserver.localhost/connect/token",
                    ClientId = "clientId",
                    ClientSecret = "clientSecret",
                    Scope = "the_scope"
                });
            });
            builder.Services.AddDistributedMemoryCache();
            builder.Services.AddTransient<ITokenClientConfigurationService, DefaultTokenClientConfigurationService>(s => 
            {
                return new DefaultTokenClientConfigurationService(
                    s.GetRequiredService<IOptions<AccessTokenManagementOptions>>(), 
                    null,
                    null);
            });
            builder.Services.AddHttpClient(AccessTokenManagementDefaults.BackChannelHttpClientName);
            builder.Services.TryAddTransient<ITokenEndpointService, TokenEndpointService>();
            builder.Services.TryAddTransient<IClientAccessTokenCache, ClientAccessTokenCache>();
            builder.Services.AddTransient<IAccessTokenManagementService, AccessTokenManagementService>(s => 
            {
                return new AccessTokenManagementService(
                    null, 
                    null, 
                    s.GetRequiredService<IOptions<AccessTokenManagementOptions>>(),
                    s.GetRequiredService<ITokenEndpointService>(),
                    s.GetRequiredService<IClientAccessTokenCache>(),
                    null,
                    s.GetRequiredService<ILogger<AccessTokenManagementService>>());
            });
            builder.Services.AddTransient<ClientAccessTokenHandler>();
            
            // End custom token client configuration.

            builder.Services.AddClientAccessTokenClient("apiClient", configureClient: config => 
            {
                config.BaseAddress = new System.Uri("http://api.localhost");
            });
        }
    }
}
```

I'm not 100% exactly which part of the default setup causes the error message:

```
An unhandled host error has occurred.
Microsoft.AspNetCore.Authentication.Core: No authentication handlers
are registered. Did you forget to call
AddAuthentication().Add[SomeAuthHandler]("WebJobsAuthLevel",...)?.
```

But it could be the dependency on `IAuthenticationSchemeProvider` in the `DefaultTokenClientConfigurationService`.

**Comment by Vidarls** [comment]
Update, Seems it was not the `IAuthenticationSchemeProvider` in `DefaultTokenClientConfigurationService` 

Tried the following config:

```c#
public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddTransient<ITokenClientConfigurationService, DefaultTokenClientConfigurationService>(s => 
            {
                return new DefaultTokenClientConfigurationService(
                    s.GetRequiredService<IOptions<AccessTokenManagementOptions>>(), 
                    null,
                    null);
            });
            builder.Services.AddAccessTokenManagement(o => 
            {
                o.Client.Clients.Add("identityserver", new IdentityModel.Client.ClientCredentialsTokenRequest
                {
                    Address = "http://myIdentityserver.local/connect/token",
                    ClientId = clientId
                    ClientSecret = "clientSecret",
                    Scope = "the_scope"
                });
            });            
            
            builder.Services.AddClientAccessTokenClient("apiClient", configureClient: config => 
            {
                config.BaseAddress = new System.Uri("http://api.localhost");
            });
        }
    }
```
And it was back to the Error message: 

```
An unhandled host error has occurred.
Microsoft.AspNetCore.Authentication.Core: No authentication handlers
are registered. Did you forget to call
AddAuthentication().Add[SomeAuthHandler]("WebJobsAuthLevel",...)?.
```

For now I'll work with this, any input from @leastprivilege would be much appreciated on how to make the config a little bit more robust at least.

**Comment by leastprivilege** [comment]
So yea - the real problem is that the even the client creds only plumbing requires the full asp.net core authentication infrastructure - which is not a problem in web and worker hosts.

It seems to be a problem for whatever customized host a function app is using.

This probably needs re-factoring. I don't have time for that right now unfortunately.

**Comment by leastprivilege** [comment]
OK - now I am confused as well. Seems only the configuration service has a dependency on the authentication handler plumbing...have you tried completely replacing the config service and read from the options yourself?

**Comment by Vidarls** [comment]
No, I have not done any more work on this yet.
My particular use case right now is for client credentials only and the above config code seems to work for that use case.
However, I have scheduled some time for refactoring that part some time during the next two weeks, so I'll have a look again then.

I was originally thinking of doing a PR adding a servicebuilder extension wrapping my new config to provide a "client credentials only" registration. But I'll look into how much work reworking the configuration service would be as well.

**Comment by leastprivilege** [comment]
I am not talking about PRs or refactoring. Simply try to implement the config service yourself and see if it solved the problem. Then we can brainstorm if/what action is needed.

**Comment by leastprivilege** [comment]
Any update on the issue? closing for now - feel free to re-open if it needs further discussion.
